### PR TITLE
Rococo Parachain XCM Adjustments

### PR DESCRIPTION
**Pull request by apopiak**
This PR:
+ adjusts the `AssetsFrom` filter to allow for suffixes longer than 1 `Junction` to be compatible with the new `PalletInstance` prefix introduced by https://github.com/paritytech/cumulus/pull/831
  + and adds some XCM tracing.
+ sets Statemint (Parachain 1000) as a trusted reserve for relay chain tokens.

Questions coming out of this:
+ Do we want to upstream this kind of matching (allowing more than 1 Junction suffix) into the XCM modules? (e.g. a `fn match_prefix(self, prefix: MultiLocation)` on `MultiLocation`? --> Yes: https://github.com/paritytech/polkadot/pull/4835

### CODE REVIEW

File: polkadot-parachains/parachains-common/src/impls.rs [hunk 1]
```diff
@@ -103,9 +103,16 @@ pub struct AssetsFrom<T>(PhantomData<T>);
...
 		let loc = T::get();
+		log::trace!(target: "xcm::AssetsFrom", "loc: {:?}, origin: {:?}", loc, origin);
 		&loc == origin &&
-			matches!(asset, MultiAsset { id: AssetId::Concrete(asset_loc), fun: Fungible(_a) }
-			if asset_loc.match_and_split(&loc).is_some())
+			match asset {
+				MultiAsset { id: AssetId::Concrete(asset_loc), fun: Fungible(_a) } =>
+					loc.parent_count() == asset_loc.parent_count() &&
+						loc.interior().iter().zip(asset_loc.interior().iter()).all(
+							|(prefix_junction, asset_junction)| prefix_junction == asset_junction,
```
Comment (gilescope) [inline polkadot-parachains/parachains-common/src/impls.rs:112]: if asset_loc.interior() is longer, then I think std iter zip will ignore any extra length. Is this intended?

Comment (apopiak) [reply polkadot-parachains/parachains-common/src/impls.rs:113]: Should be fixed now.

File: polkadot-parachains/parachains-common/src/impls.rs [hunk 2]
```diff
@@ -98,14 +98,41 @@ where
...
+/// ```
+/// # use xcm::prelude::*;
+/// # use parachains_common::impls::match_prefix;
+/// let prefix =  MultiLocation::new(1, X1(Parachain(1234)));
+/// let loc =  MultiLocation::new(1, X2(Parachain(1234), GeneralIndex(5)));
+/// assert!(match_prefix(&prefix, &loc));
+/// let failing =  MultiLocation::new(1, Here);
+/// assert!(!match_prefix(&prefix, &failing));
+/// ```
+pub fn match_prefix(prefix: &MultiLocation, loc: &MultiLocation) -> bool {
```
Comment (KiChjang) [inline polkadot-parachains/parachains-common/src/impls.rs:114]: This looks like it's a function that should be upstreamed to Polkadot.

Comment (apopiak) [reply polkadot-parachains/parachains-common/src/impls.rs:114]: yep, agree, as mentioned in the description :-D

Comment (apopiak) [reply polkadot-parachains/parachains-common/src/impls.rs:114]: https://github.com/paritytech/polkadot/pull/4827

File: polkadot-parachains/parachains-common/src/impls.rs [hunk 3]
```diff
@@ -98,14 +98,33 @@ where
 	}
 }
 
+/// Tests `loc` whether it starts with `prefix`.
+/// Similar to `MultiLocation::match_and_split` but allows arbitrary suffixes (instead of just one
+/// junction).
+// TODO: Replace with version implemented on MultiLocation here:
+// https://github.com/paritytech/polkadot/pull/4827
```
Comment (apopiak) [inline polkadot-parachains/parachains-common/src/impls.rs:105]: XCM v3 might take a while to be merged. I can just create an issue to track this.

**Review by KiChjang — APPROVED** [review]

**Comment by joepetrowski** [comment]
@KiChjang @NachoPal : Is this still a change we want/need? Let's merge master or close it.

**Comment by KiChjang** [comment]
IIRC, the discussion we had before was that Statemint/e shouldn't view the relay chain as a reserve for DOT/KSM as it can cause problems with reserves drainage via the mixed usage of asset teleportation and reserve transfers. Given so, we probably do not need to allow Statemint to recognize the relay chain as a reserve.

However, the `AssetsFrom` filter change still makes sense to me.

**Pull request closed without merging**